tokio/src/fs/read_uring.rs

```diff
@@ -0,0 +1,183 @@
+use crate::fs::OpenOptions;
+use crate::runtime::driver::op::Op;
+
+use std::io::ErrorKind;
+use std::os::fd::OwnedFd;
+use std::path::Path;
+use std::{cmp, io};
+
+// this algorithm is inspired from rust std lib version 1.90.0
+// https://doc.rust-lang.org/1.90.0/src/std/io/mod.rs.html#409
+const PROBE_SIZE: usize = 32;
+const PROBE_SIZE_U32: u32 = PROBE_SIZE as u32;
+
+// Max bytes we can read using io uring submission at a time
+// SAFETY: cannot be higher than u32::MAX for safe cast
+const MAX_READ_SIZE: usize = u32::MAX as usize;
+
+pub(crate) async fn read_uring(path: &Path) -> io::Result<Vec<u8>> {
+    let file = OpenOptions::new().read(true).open(path).await?;
+
+    // TODO: use io uring in the future to obtain metadata
+    let size_hint: Option<usize> = file
+        .metadata()
+        .await
+        .map(|m| usize::try_from(m.len()).unwrap_or(usize::MAX))
+        .ok();
+
+    let fd: OwnedFd = file
+        .try_into_std()
+        .expect("unexpected in-flight operation detected")
+        .into();
+
+    // extra single capacity for the whole size to fit without any reallocation
+    let buf = Vec::with_capacity(size_hint.unwrap_or(0).saturating_add(1));
```

Comment:
> extra single capacity for the whole size to fit without any reallocation

Why doesn't the standard library have this issue?

```rust
// https://doc.rust-lang.org/1.90.0/src/std/fs.rs.html#308
pub fn read<P: AsRef<Path>>(path: P) -> io::Result<Vec<u8>> {
    fn inner(path: &Path) -> io::Result<Vec<u8>> {
        let mut file = File::open(path)?;
        let size = file.metadata().map(|m| m.len() as usize).ok();
        let mut bytes = Vec::try_with_capacity(size.unwrap_or(0))?;
        io::default_read_to_end(&mut file, &mut bytes, size)?;
        Ok(bytes)
    }
    inner(path.as_ref())
}
```

Reply:
Because they react on the bytes read rather than how many bytes to read ahead of time. In case of a short read during EOF and try_reserve is called, it allocates more space than needed and also doubles the capacity but if its +1 then there's no need and capacity isn't necessarily grown